### PR TITLE
#151844125 Get all available roles.

### PR DESCRIPTION
**Pull request by sirghiny**
#### What does this PR do?

- Implement get all roles available roles.

#### Description of Task to be completed?

- Create endpoint get api/v1/roles/
- Write tests for endpoint

#### How should this be manually tested?

`pytest -V` or `pytest`

#### What are the relevant pivotal tracker stories?

151844125

### CODE REVIEW

**Comment by Mnickii** [comment]
Please rebase

**Comment by andela-ojoloko** [comment]
please rebase and squash your commits into 1

**Comment by Mnickii** [comment]
Your PR message deviates from team conventions

File: api/views/roles.py [hunk 1]
```diff
@@ -32,3 +32,24 @@ def get(self, role_id):
...
                     "message": "The specified role does not exist."}
             }, 404
+
+class RolesResource(Resource):
+    """Contain GET for all available roles."""
+    @token_required
+    def get(self):
+        roles = Role.fetch_all()
+        if roles:
+            roles_list = [role.name for role in roles]
```
Comment (andela-ojoloko) [inline api/views/roles.py]: please send the entire role object not just the name. you should do `roles.serialize` rather than `role.name` in the list comprehension.

Comment (sirghiny) [reply api/views/roles.py]: Okay, I'll see to that.

File: test/api/views/test_roles.py [hunk 2]
```diff
@@ -132,3 +132,16 @@ def test_users_from_empty_role(self):
...
+        """
+        response = self.client.get('/api/v1/roles/',
+                                   headers=self.regular_header)
+        response_data = json.loads(response.data)["data"]["roles"]
+        expected = ["Leaning Map User",
+                    "Leaning Map Administrator",
+                    "Leaning Map Super Administrator",
+                    "Learning Map No Users"]
+        self.assertEqual(expected, response_data)
+        self.assert200(response)
```
Comment (andela-ojoloko) [inline test/api/views/test_roles.py]: please write tests for all edge cases like when they are no roles available, the request contains no authorization header etc

File: api/views/roles.py [hunk 3]
```diff
@@ -32,3 +32,24 @@ def get(self, role_id):
...
+            return {
+                "status": "success",
+                "data": {
+                    "roles": roles_list
+                    }
+            }
+        else:
```
Comment (andela-milesanmi) [inline api/views/roles.py]: Is the else statement here needed? I believe that if the if condition fails, the rest of the code should run automatically.

File: api/views/roles.py [hunk 4]
```diff
@@ -32,3 +32,24 @@ def get(self, role_id):
...
+            return {
+                "status": "fail",
+                "data": {
+                    "message": "There are currently no roles available."
+                    }
+                }
```
Comment (andela-milesanmi) [inline api/views/roles.py]: What `status_code` do we have here?

File: api/views/roles.py [hunk 5]
```diff
@@ -32,3 +32,24 @@ def get(self, role_id):
...
+    def get(self):
+        roles = Role.fetch_all()
+        if roles:
+            roles_list = [role.name for role in roles]
+            return {
+                "status": "success",
+                "data": {
+                    "roles": roles_list
+                    }
+            }
```
Comment (andela-milesanmi) [inline api/views/roles.py]: What is the `status_code`?

File: main.py [hunk 6]
```diff
@@ -152,8 +152,11 @@ def index():
     api.add_resource(RoleUsersResource,  
                      '/api/v1/roles/<string:role_id>/users',
                      '/api/v1/roles/<string:role_id>/users/',
-                      endpoint='role_users'
-    )
+                     endpoint='role_users')
+
+    api.add_resource(RolesResource,
+                     '/api/v1/roles/',
+                     endpoint='roles')
```
Comment (wanjiku-carol) [inline main.py]: add in the endpoint the one for `'/api/v1/roles'` as well

File: api/views/roles.py (hunk 3 above)
Comment (sirghiny) [reply api/views/roles.py]: It is not necessary. I had included it just for readability.

File: api/views/roles.py [hunk 7]
```diff
@@ -20,15 +20,34 @@ def get(self, role_id):
...
+        roles = Role.fetch_all()
+        if roles:
+            roles_list = [role.serialize() for role in roles]
+            return {
+                "status": "success",
+                "data": {
+                    "roles": roles_list
+                    }
+            }, 200
+        return {
```
Comment (andela-ojoloko) [inline api/views/roles.py]: It might be better to add the else statement back to make the code more readable. code `if else` statements for checking whether values exist are common through out the learning map api codebase. It makes the code more easy to understand.

**Comment by andela-ojoloko** [comment]
LGTM

File: api/views/roles.py (hunk 7 above)
Comment (andela-milesanmi) [reply api/views/roles.py]: I believe that readability is one of the primary purposes of Python's indentation. Hence, it is not necessary.

**Comment by wanjiku-carol** [comment]
LGTM

**Comment by Mnickii** [comment]
LGTM
